### PR TITLE
Added next_up and next_down for f32/f64.

### PR DESCRIPTION
**Pull request by orlp**
This is a pull request implementing the features described at https://github.com/rust-lang/rfcs/pull/3173.

### CODE REVIEW

**Comment by kpreid** [comment]
In the RFC you mention “Repeatedly calling next_up on a negative number will iterate over all values above it, with the exception of +0.0, only -0.0 will be visited.” I think this point would be good to make explicitly in the documentation of these functions — it is a place where the two goals “exhaust the space of float-representable numbers” and “move along the number line in the smallest nonzero step” differ, and it's worth mentioning this to, if nothing else, prompt the programmer to think about which thing they're doing and whether the function's inherent behavior is correct for their use case.

**Comment by scottmcm** [comment]
:+1: to having these.  I think we should explicitly mention that these are expected to match IEEE's `nextUp` and `nextDown` operations, and keep the `next_up` and `next_down` names.

(I don't think we'll find any name that's substantially better, so sticking with precedent seems best.)

**Comment by orlp** [comment]
So investigating why my pull request failed, I'm not sure that it's my fault. This fails on i586:

```rust
f32::from_bits(0x7F955555).to_bits() == 0x7F955555
```

This violates the standard library guarantees, which say that `from_bits` and `to_bits` are transmutes.

**Comment by scottmcm** [comment]
NANs are always a bit weird, and pre-i686 floats are particularly annoying.  You might be interested in the conversations in https://github.com/rust-lang/rust/issues/73328

But honestly for now I'd just say `cfg` out that test on i586.

**Comment by orlp** [comment]
@scottmcm I managed to rewrite my test case to simply avoid the issue altogether I believe by converting to `f32` earlier:

```rust
// Check that NaNs roundtrip.
let nan0 = f32::NAN;
let nan1 = f32::from_bits(f32::NAN.to_bits() ^ 0x002a_aaaa);
let nan2 = f32::from_bits(f32::NAN.to_bits() ^ 0x0055_5555);
assert_eq!(nan0.next_down().to_bits(), nan0.to_bits());
assert_eq!(nan1.next_down().to_bits(), nan1.to_bits());
assert_eq!(nan2.next_down().to_bits(), nan2.to_bits());
```

Now the correctness of the testcase does not rely on `f32::from_bits(u).to_bits() == u`.

I also rebased to resolve the merge conflicts so this should be ready to merge once again.

**Comment by orlp** [comment]
I am *really* puzzled how this could fail. Note the test case that fails:

```rust
let nan0 = f32::NAN;
let nan1 = f32::from_bits(f32::NAN.to_bits() ^ 0x002a_aaaa);
let nan2 = f32::from_bits(f32::NAN.to_bits() ^ 0x0055_5555);
assert_eq!(nan0.next_down().to_bits(), nan0.to_bits());
assert_eq!(nan1.next_down().to_bits(), nan1.to_bits());
assert_eq!(nan2.next_down().to_bits(), nan2.to_bits()); // Fails.
```

And the implementation:

```rust
pub const fn next_down(self) -> Self {
    // Ommitted for brevity...
    if self.is_nan() || bits == Self::INFINITY.to_bits() {
        return self;
    }
    // Ommitted for brevity...
```

Note that we actually just return `self` when `self` is a nan. So how can `nan2.next_down().to_bits() == nan2.to_bits()` fail?

I can only imagine four scenarios:

1. `f32::from_bits(f32::NAN.to_bits() ^ 0x0055_5555)` somehow does not form a nan value even though it's only modifying the odd mantissa bits of what used to be a nan value.
2. `f32::is_nan` is bugged.
3. `f32::to_bits` is non-deterministic.
4. A compiler bug.

I have no idea how to get access to a `dist-i586-gnu-i586-i686-musl` platform to reproduce for some investigation to find which of these scenarios occurs. The testcases pass on the `i586-pc-windows-msvc` and `i686-pc-windows-msvc` targets.

For those who want a stand-alone testcase, see this playground: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=737192093518dcc3a3d3e107dd5f2a31

**Comment by orlp** [comment]
There's a fifth scenario @SkiFire13 suggests as a possibility: the floating point value `nan2` canonicalizes when passed to `next_down` because it has to be passed on the stack, but does not do so when passed directly to `to_bits`.

**Comment by SkiFire13** [comment]
I think the issue is the same as #73288, except now debug mode is not required, probably due to LLVM not inlining `nan2.next_down()`.

> because it has to be passed on the stack

It has to be passed on the FPU stack, which means it has to become extended-precision and then truncated back to 32 bits. Since the latter is an arithmetic operation, it has to be canonicalized. See https://github.com/rust-lang/rust/issues/73328#issuecomment-692186924

**Comment by scottmcm** [comment]
@bors r-

Again, the easiest option here might just be to skip these on i586, since there's lots of float stuff that doesn't work the same there.

**Comment by orlp** [comment]
@scottmcm Do you have a suggested way of checking for that platform?

**Comment by scottmcm** [comment]
Hmm, I was expecting something like `cfg!(target_cpu = "i586")`, but it looks like maybe that doesn't exist? <https://doc.rust-lang.org/reference/attributes/codegen.html#the-target_feature-attribute>

Maybe you can fake it with `x86`-but-not-`sse` or something.

File: library/std/src/f32/tests.rs [hunk 1]
```diff
@@ -287,6 +287,69 @@ fn test_is_sign_negative() {
     assert!((-f32::NAN).is_sign_negative());
 }
 
+#[test]
```
Comment (Urgau) [inline library/std/src/f32/tests.rs:290]: Maybe this ? This seems enough to [only exclude](https://godbolt.org/z/KarxqzEna) `i586-unknown-linux-gnu`.
```suggestion
#[test]
#[cfg_attr(not(all(target_arch = "x86", not(target_feature = "fxsr"))), ignore)]
```

Comment (orlp) [reply library/std/src/f32/tests.rs:290]: What is the `fxsr` target feature? Either way I would only exclude the NaN block from the test, not the whole test.

Comment (Urgau) [reply library/std/src/f32/tests.rs:290]: From what I can tell this feature tell if these float `FXSAVE, FXRESTOR` instructions are available.
This CPU feature is not be available for the `pentium` cpu which is the base cpu of `i586-unknown-linux-gnu` but is available after that. This was the only difference I could find to only exclude the `i586-unknown-linux-gnu` target.

**Comment by orlp** [comment]
I'm going to try one more thing before going into conditional compilation to see if it works: forcibly pass the reference NaN value through a function boundary. Sadly I can't test locally as I don't have access to a i586 platform, so I'll have to rely on the CI.

If this doesn't work I'll give @Urgau 's idea a spin.

**Comment by scottmcm** [comment]
Let's give it a try
@bors r=yaahc,scottmcm

File: library/core/src/num/f32.rs [hunk 2]
```diff
@@ -620,6 +620,104 @@ impl f32 {
...
+    /// // But not for most numbers.
+    /// assert!(0.1f32.next_up() < 0.1 + f32::EPSILON);
+    /// assert_eq!(16777216f32.next_up(), 16777218.0);
+    /// ```
+    ///
+    /// [`NEG_INFINITY`]: Self::NEG_INFINITY
+    /// [`INFINITY`]: Self::INFINITY
+    /// [`MIN`]: Self::MIN
+    /// [`MAX`]: Self::MAX
+    #[unstable(feature = "float_next_up_down", issue = "none")]
```
Comment (Urgau) [inline library/core/src/num/f32.rs:649]: All of the `[unstable]` attributes are missing a tracking issue. It can be created by following the template [here](https://github.com/rust-lang/rust/issues/new?assignees=&labels=C-tracking-issue%2C+T-libs-api&template=library_tracking_issue.md&title=Tracking+Issue+for+XXX).

File: library/core/src/num/f32.rs [hunk 3]
```diff
@@ -620,6 +620,104 @@ impl f32 {
...
+    pub const fn next_up(self) -> Self {
```
Comment (Urgau) [inline library/core/src/num/f32.rs:650]: They are also all missing the `[rustc_const_unstable]` attribute (which is not currently enforced due to a bug in rustc).
```suggestion
    #[rustc_const_unstable(feature = "float_next_up_down", issue = "?")]
    pub const fn next_up(self) -> Self {
```

**Comment by yaahc** [comment]
I've gone ahead and merged the RFC and setup a tracking issue, https://github.com/rust-lang/rust/issues/91399. Could you update the unstable attributes to refer to this issue?

**Comment by orlp** [comment]
I can. And it seems my attempted fix was not successful... What a nasty platform that `i586` is. Oh well, conditional compilation it is.

File: library/core/src/num/f32.rs (hunk 3 above)
Comment (orlp) [reply library/core/src/num/f32.rs:650]: Had no idea this was a thing, will add them.

File: library/core/src/num/f32.rs (hunk 2 above)
Comment (orlp) [reply library/core/src/num/f32.rs:649]: yaahc created one, will refer to them.

File: library/std/src/f32/tests.rs (hunk 1 above)
Comment (orlp) [reply library/std/src/f32/tests.rs:290]: I don't want to skip the entire test case on this platform, only the NaN round-trip, so I'm just going to conditionally compile out the NaN payload block only.

**Comment by orlp** [comment]
@scottmcm I've given up trying to cleverly construct the test around i586, so I've just added conditional compilation now. If @Urgau 's suggested config works it should pass all tests.

**Comment by scottmcm** [comment]
🤞

@bors r=yaahc,scottmcm

**Comment by Urgau** [comment]
> ... I don't understand `dist-i586-gnu-i586-i686-musl`. I guess it also _sometimes_ flushes denormals, but not on the reference values. Probably something to do with normalization only on function boundaries again.

Btw, my `cfg` block only covers `i586-unknown-linux-musl` but the tests failed on `i686-unknown-linux-musl`. If you want to also ignore that one: `#[cfg(not(target_arch = "x86"))]` should do the trick.

> At this point I am thinking of just skipping the whole platform in the tests entirely.

If you do that, don't forget to add this to the tracking issue as an "Unresolved Questions".

> I'd test a bunch of things but since I don't have access to the platform the feedback loop through CI is... not ideal.

You could maybe try run the tests (from you own binary or directly from std) with `qemu-i386-static` from `qemu-user-static` (in the repo of all major distributions).

**Comment by matthiaskrgr** [comment]
Still failing in ci I guess: https://github.com/rust-lang/rust/pull/91983#issuecomment-995379026

@bors rollup=never r-

**Comment by orlp** [comment]
I don't know how it was marked to be ready to merged since we knew it was failing two weeks ago, sorry.

The problem is that x87 floating point is very problematic for this set of functions. At first I thought it was just NaN bits not being preserved, but it appears that it's also flushing denormals to zero, not just when doing arithmetic operations, but also *when passing values through function boundaries*.

I think I'll disable testing on x87 floating point platforms for now and indeed add an unresolved question to the tracking issue. I honestly don't know what to do with this platform.

**Comment by matthiaskrgr** [comment]
Oh strange, you're right I didn't see that.
Sometimes bors acts funny and the queue shows prs that were already unapproved as approved, guess that was one of those times.
Sorry for the noise... :smile:

File: library/core/src/num/f32.rs [hunk 4]
```diff
@@ -620,6 +620,106 @@ impl f32 {
...
+    ///
+    /// Let `TINY` be the smallest representable positive `f32`. Then,
+    ///  - if `self.is_nan()`, this returns `self`;
+    ///  - if `self` is [`NEG_INFINITY`], this returns [`MIN`];
+    ///  - if `self` is `-TINY`, this returns -0.0;
+    ///  - if `self` is -0.0 or +0.0, this returns `TINY`;
+    ///  - if `self` is [`MAX`] or [`INFINITY`], this returns [`INFINITY`];
+    ///  - otherwise the unique least value greater than `self` is returned.
+    ///
+    /// The identity `x.next_up() == -(-x).next_down()` holds for all `x`. When `x`
```
Comment (goffrie) [inline library/core/src/num/f32.rs:633]: This is technically not true for `x = NAN`.

Comment (orlp) [reply library/core/src/num/f32.rs:633]: Why not?

Comment (orlp) [reply library/core/src/num/f32.rs:633]: Oh I think I see what you're getting at, `==` here is to be understood as a bitwise / definition equality, not floating point comparison where `NaN != NaN`.

**Comment by JohnCSimon** [comment]
ping from triage:
@orlp Can you address the review comments and resolve the conversation? we have some build failures, too. 

When it's ready for review send a message containing 
`@rustbot ready` to switch to `S-waiting-on-review`

**Comment by orlp** [comment]
@[Dylan-DPC](https://github.com/Dylan-DPC) I think I've been procrastinating on this because I don't see a good solution. Today I'll update the tracking issue with the unresolved question on what to do with x87, disable the testing for that platform for now and push my changes.

**Comment by orlp** [comment]
I've posted the unresolved question to the tracking issue: https://github.com/rust-lang/rust/issues/91399#issuecomment-1045989643

Should I wait with asking a review for this until the unresolved question is resolved or do it now @JohnCSimon ?

File: library/core/src/num/f32.rs (hunk 4 above)
Comment (orlp) [reply library/core/src/num/f32.rs:633]: Should I update this doc comment to say "for all non-NaN `x`"? Or maybe reword it to "`x.next_up()` is equivalent to `-(-x).next_down()` for all `x`" to prevent confusion with `==` failing for NaNs?

**Comment by JohnCSimon** [comment]
@orlp - if you think it's ready for review, go ahead and mark it ready for review. I'm just doing PR triage.

**Comment by JohnCSimon** [comment]
@orlp 
Ping from triage: I'm closing this due to inactivity, Please reopen when you are ready to continue with this. 
Note: if you do please open the PR **BEFORE** you push to it, else you won't be able to reopen - this is a quirk of github. 
Thanks for your contribution.

@rustbot label: +S-inactive
